daemon: allow setting OCI annotations on containers

corhere opened the pull request:
- Relates to #45030
- Closes #39142
- Closes #37289
- Closes #36181
- Relates to #37262

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

Allow API clients to set and retrieve OCI annotations on a container.

**- How I did it**

Added a field to `HostConfig` and plumbed it into the OCI spec generation code.

**- How to verify it**
Create a container with annotations and start it. Inspect the running container using `ctr` and verify that the annotations are listed in the container spec.

A new integration test verifies that annotations can be round-tripped through the Engine API.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
- Annotations—arbitrary non-identifying metadata—can now be attached to containers on creation which will be passed to the runtime when the container is started.

**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah commented:
Will these labels be committed in the image on `docker commit` ? (also see https://github.com/moby/moby/pull/42037) And does this mean that an image containing these labels will set them at runtime on the container?

corhere commented:
> Will these labels be committed in the image on `docker commit` ? (also see #42037)

Yes, until the aforementioned PR is merged.

> And does this mean that an image containing these labels will set them at runtime on the container?

Yes. This could arguably be a feature.

thaJeztah commented:
> Yes. This could arguably be a feature.

We should probably look closely what's desirable in that case. I recall there was some prior discussion on passing through labels as annotations. I must admit I'm not fully aware how they're used / what they're able to control, but ISTR there were some concerns that had to be looked into (e.g. a malicious image could set a label and do other things than expected).

I guess to some extent that's not "our" issue (as we don't consume those annotations?).

It does somewhat bring back the issue though that we don't have a concept of "runtime" labels and "non-runtime" (image labels are runtime labels in our case, or at least inherited directly, and in case of "commit", baked into the image).

Probably also need to be looked at what the expectations are for `docker container update`, as there's been quite some requests on being able to update labels; prior discussions stranded on that one, but if that would be implemented, that may have an effect on the feature in this PR.

AkihiroSuda reviewed (changes requested):
> > And does this mean that an image containing these labels will set them at runtime on the container?
>
> Yes. This could arguably be a feature.

This looks like an attack vector.

Specifying annotations should require a new API/CLI like 
`docker run --annotation=com.example.foo=bar`.
(similar to `podman run --annotation`)

corhere committed:
daemon: add annotations to container HostConfig

Allow clients to set annotations on a container which will applied to
the container's OCI spec.

Signed-off-by: Cory Snider <csnider@mirantis.com>

cpuguy83 reviewed (changes requested):
SGTM

But need to update the API history doc and openapi yaml.

AkihiroSuda commented on api/types/container/hostconfig.go:
```diff
@@ -377,16 +377,17 @@ type UpdateConfig struct {
+	VolumeDriver    string            // Name of the volume driver used to mount volumes
+	VolumesFrom     []string          // List of volumes to take from other container
+	ConsoleSize     [2]uint           // Initial console size (height,width)
+	OCIAnnotations  map[string]string `json:",omitempty"` // Annotations to pass along to OCI runtime
```
Why `json:",omitempty"` ?

corhere committed:
docs/api: add v1.43 OpenAPI definitions

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere commented on api/server/router/container/container_routes.go:
```diff
@@ -563,6 +563,11 @@ func (s *containerRouter) postContainersCreate(ctx context.Context, w http.Respo
 		hostConfig.ConsoleSize = [2]uint{0, 0}
 	}
 
+	if hostConfig != nil && versions.LessThan(version, "1.43") {
+		// Ignore OCIAnnotations because it was added in API v1.43.
+		hostConfig.OCIAnnotations = nil
+	}
```
I'm following past precedent by ignoring the field on create for older API versions but unconditionally including it in the `/containers/{}/json` response for all API versions.

thaJeztah commented on api/types/container/hostconfig.go:
```diff
@@ -377,16 +377,17 @@ type UpdateConfig struct {
+	VolumeDriver    string            // Name of the volume driver used to mount volumes
+	VolumesFrom     []string          // List of volumes to take from other container
+	ConsoleSize     [2]uint           // Initial console size (height,width)
+	OCIAnnotations  map[string]string `json:",omitempty"` // Annotations to pass along to OCI runtime
```
Before we merge; quick check on the naming. Wondering if `OCI` prefix is needed in our API here. I get where we're coming from; mostly thinking along the lines of "(almost) every option we have in our API leads to `<something>` OCI spec.

So wondering if we should keep the abstraction - call it `Annotation`, and keep the fact that that (currently) gets passed on to the OCI to the description only.

corhere replied:
I'm coming around to dropping the `OCI` prefix and adopting essentially [the same semantics as Kubernetes `Annotations`.](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/) And by that I mean documenting that Engine API clients can also leverage annotations to attach and consume metadata on containers for their own purposes. Perhaps we should also follow their lead and [enforce a particular syntax for annotation keys](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#syntax-and-character-set) so that the namespace of an annotation is unambiguous. We could always loosen validation later but can't make it more restrictive.

thaJeztah replied:
Let's discuss / brainstorm in the call later Today (I also need to get my head around all of that). Quick comment;

- Yes, I think Annotations could be useful for "our side" as well
- I like namespaces / prefixes as they can definitely help prevent ambiguity (I think I may actually have been the one originally getting them into the "labels" definition 🤔 )
- ^^^ w.r.t. namespaces: I always keep in the back of my mind that `API !== UX`; namespaces _are_ awkward to use, but are (in most cases) not meant for humans. So if we decide to use namespaces, we can still make the UX "different" (`docker run --foo=bar` doing `docker run --annotation=awkwardly-long-namespace.option=bar` behind the scenes)

corhere replied:
What I like about how K8s goes about namespacing annotations is that it makes namespace prefixes optional and [reserves the no-namespace "namespace" for users.](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#syntax-and-character-set:~:text=If%20the%20prefix%20is%20omitted%2C%20the%20annotation%20Key%20is%20presumed%20to%20be%20private%20to%20the%20user.)

thaJeztah replied:
Yup, that was part of my original suggestion for labels; no-namespace is for users; don't use them for automation as they can easily conflict. For automation / tools, always use a prefix

thaJeztah replied:
For those interested; here's the PR with all the fun (well, there's been multiple before that, but that's the one that made it, and that one is already 200+ comments 🤣)

- https://github.com/moby/moby/pull/9882

corhere replied:
Decisions from today's maintainer call:
- Drop the `OCI` prefix from the field name
- Any annotation which is valid per the OCI Runtime spec is permitted
  - Keys can be any JSON string except for the empty string
  - Values can be any JSON string, including the empty string
  - Keys and values can be arbitrarily long

thaJeztah reviewed (changes requested):
("requesting changes" just to have a quick check per my comment - changes otherwise look good to me)

thaJeztah commented on daemon/container.go:
```diff
@@ -305,6 +305,11 @@ func validateHostConfig(hostConfig *containertypes.HostConfig) error {
 	}
+	for k := range hostConfig.Annotations {
+		if k == "" {
+			return errors.Errorf("invalid Annotations: the empty string is not permitted as an annotation key")
```
I had a look if we should be using an `errdefs.InvalidParameter()` here (and noticed other errors didn't), but it looks like we wrap any errors returned here on the call sites, such as https://github.com/moby/moby/blob/2f0e308c7d4cb59f33664b763fa95542dee3c024/daemon/create.go#L65-L68


So in either case; it's good (currently) to not use an errdefs, but we _could_ (should?) look at these at some points to assign error-types in these locations.

thaJeztah reviewed (approved):
LGTM, thanks!

thaJeztah commented:
@cpuguy83 LGTY? (I see there's a "change requested" from you, and didn't want to dismiss it)

<img width="506" alt="Screenshot 2023-02-24 at 14 56 38" src="https://user-images.githubusercontent.com/1804568/221196359-179bdedf-5ce0-4fe5-9e7e-56a0ad77efb4.png">

AkihiroSuda commented:
Opened a PR for the CLI:
- https://github.com/docker/cli/pull/4156
